### PR TITLE
jar squishing, part 8

### PR DESCRIPTION
**Pull request by rhysdh540**
- [x] make all mixin classes abstract (just good practice)
- [x] apply the refmap to mixin classes at compile time, effectively rendering it useless (and then I removed it)
- [x] migrate to a new JSON5 library, ZSON
  - [x] merge https://github.com/Nolij/ZSON/pull/1
  - [x] add versioning and maven publishing
  - [x] publish to maven
 - [x] remove a bunch of classes that could be replaced
 - [x] obfuscate Fabric entrypoints
 - [x] repackage mixin classes to `zume.mixin` (optional)
   - [ ] obfuscate mixin class names

### CODE REVIEW

File: api/src/main/java/dev/nolij/zume/impl/config/ZumeConfigImpl.java
Comment (Nolij) [inline api/src/main/java/dev/nolij/zume/impl/config/ZumeConfigImpl.java:82]: Please keep this as a multi-line comment for readability.

File: api/src/main/java/dev/nolij/zume/impl/config/ZumeConfigImpl.java
Comment (Nolij) [inline api/src/main/java/dev/nolij/zume/impl/config/ZumeConfigImpl.java:202]: I feel like this belongs in ZSON.

File: archaic/src/main/java/dev/nolij/zume/mixin/archaic/EntityRendererMixin.java
Comment (Nolij) [inline archaic/src/main/java/dev/nolij/zume/mixin/archaic/EntityRendererMixin.java:13]: Why?

File: lexforge/build.gradle.kts
Comment (Nolij) [inline lexforge/build.gradle.kts:20]: This can probably be moved to the main `build.gradle.kts` with an `implName in lexForgeImpls` check or something like that.

File: build.gradle.kts
Comment (Nolij) [inline build.gradle.kts:173]: This should probably be removed before merging.

File: build.gradle.kts
Comment (Nolij) [inline build.gradle.kts:365]: This will almost definitely remove the `@Mod` from the LexForge entrypoint. That's why I explicitly listed the target classes.

File: build.gradle.kts
Comment (Nolij) [inline build.gradle.kts:381]: `dev.nolij.zume.zson`

File: build.gradle.kts
Comment (Nolij) [inline build.gradle.kts:542]: This is why inferred typing is not an excuse for turing-style typing specification.

File: gradle.properties
Comment (Nolij) [inline gradle.properties:94]: https://github.com/Nolij/ZSON/releases/latest

File: proguard.pro
Comment (Nolij) [inline proguard.pro:49]: Please ensure class members being renamed doesn't cause bugs.

**Review by Nolij — CHANGES_REQUESTED** [review]

File: archaic/src/main/java/dev/nolij/zume/mixin/archaic/EntityRendererMixin.java
Comment (rhysdh540) [reply archaic/src/main/java/dev/nolij/zume/mixin/archaic/EntityRendererMixin.java:13]: https://github.com/SpongePowered/Mixin/wiki/Introduction-to-Mixins---Understanding-Mixin-Architecture#4-only-you-mixins-can-save-mankind

> The mixin class is marked with the `abstract` modifier. Whilst this is not a requirement, it helps when using mixins within an IDE because it means the end user cannot write code which tries to instantiate a mixin class, which would lead to an error at runtime. It also removes the requirement (imposed by the Java compiler) to implement every method within any declared interfaces, which is one of the main goals of mixins.

File: build.gradle.kts
Comment (rhysdh540) [reply build.gradle.kts:542]: the linter broke on the `!= null` check 2 lines before this and adding this fixed it
my kotlin may just be broken though

File: build.gradle.kts
Comment (rhysdh540) [reply build.gradle.kts:173]: is it bad in general to have it here? I feel like it would be fine to keep

File: build.gradle.kts
Comment (Nolij) [reply build.gradle.kts:550]: My point was turing-style typing likes to use the excuse of "you usually don't need to specify types anyway". Anyway not your doing it's Kotlin's problem.

File: build.gradle.kts
Comment (Nolij) [reply build.gradle.kts:173]: I'd prefer it not be in master. You can keep it locally just don't commit it.

File: build.gradle.kts
Comment (Nolij) [inline build.gradle.kts:174]: Please remove this. It's not about the scope, it's about it not being something that should be present in the `master` branch. You can keep it locally but it shouldn't be committed.

File: build.gradle.kts
Comment (Nolij) [reply build.gradle.kts:401]: Please restore the old whitelist approach. I don't see why a blacklist approach is an improvement.

File: lexforge/build.gradle.kts
Comment (Nolij) [inline lexforge/build.gradle.kts:47]: Might need to add ZSON here. Not sure.

File: lexforge16/build.gradle.kts
Comment (Nolij) [inline lexforge16/build.gradle.kts:19]: Same here.

File: lexforge18/build.gradle.kts
Comment (Nolij) [inline lexforge18/build.gradle.kts:20]: And here.

File: neoforge/build.gradle.kts
Comment (Nolij) [inline neoforge/build.gradle.kts:37]: E

**Review by Nolij — CHANGES_REQUESTED** [review]

File: build.gradle.kts
Comment (rhysdh540) [reply build.gradle.kts:401]: I would say that it's an improvement because it allows more lexforge entrypoints to be added in this manner without explicitly adding them to the whitelist. Correct me if I'm wrong, but I don't think that anything will need to be added to the blacklist. The only minor downside here is that it takes slightly longer to read and write every class file, but that shouldn't be a problem.

File: lexforge/build.gradle.kts
Comment (rhysdh540) [reply lexforge/build.gradle.kts:47]: this is declared in the root script.

File: build.gradle.kts
Comment (Nolij) [reply build.gradle.kts:396]: I prefer the whitelist approach. Please revert back to it.

File: proguard.pro
Comment (rhysdh540) [reply proguard.pro]: seems to work for me (I only tested in-game on modern, but the game loaded on primitve too)

File: proguard.pro
Comment (Nolij) [inline proguard.pro:15]: This doesn't error? Last time I tried to add a `**` filter to ProGuard it caused errors...

File: proguard.pro
Comment (Nolij) [reply proguard.pro]: Please also check that the keybinds register in Primitive, and that the Mod Menu integration actually works in Modern (ie that the config button is registered and opens config file on click).

File: proguard.pro
Comment (rhysdh540) [reply proguard.pro:15]: seems to work, both locally and in CI

File: buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt
Comment (Nolij) [inline buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt:303]: Please cache the `mappingTree` value.

**Review by Nolij — DISMISSED** [review]

File: proguard.pro
Comment (rhysdh540) [reply proguard.pro]: Mod(ern) Menu integration works fine. I can't test primitive because macOS just won't launch fabric, but given that MM works I would assume that it does too.

File: buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt
Comment (Nolij) [inline buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt:70]: Not sure how `null` is a cached value. I more meant to use whatever the Kotlin equivalent of `computeIfAbsent` is...

**Comment by embeddedt** [comment]
I have read and hereby affirm the entire contents of the Developer Certificate of Origin.

**Comment by Nolij** [comment]
Please wait for my review on the reformat before merging. I've found some changes the IntelliJ formatter makes are ones I disagree with and theres no way to change them in `.editorconfig` (that I've found).

File: api/src/main/java/dev/nolij/zume/api/platform/v1/IZumeImplementation.java
Comment (Nolij) [inline api/src/main/java/dev/nolij/zume/api/platform/v1/IZumeImplementation.java:8]: Please just revert the formatting changes for this file.

File: api/src/main/java/dev/nolij/zume/api/util/v1/MethodHandleHelper.java
Comment (Nolij) [inline api/src/main/java/dev/nolij/zume/api/util/v1/MethodHandleHelper.java:46]: Revert

File: build.gradle.kts
Comment (Nolij) [inline build.gradle.kts:4]: No `*` imports please.

File: buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt
Comment (Nolij) [inline buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt:10]: `*` import begone

File: buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt
Comment (Nolij) [inline buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt:22]: `*` import begone

File: buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt
Comment (Nolij) [inline buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt:88]: Why does this line wrap but the line above doesn't? Maybe make them match.

File: buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt
Comment (Nolij) [inline buildSrc/src/main/kotlin/dev/nolij/zumegradle/JarCompressing.kt:34]: Why is it removing empty line indents in Kotlin but adding them in Java? Please update `.editorconfig` so they match. I prefer indents on empty lines but prioritize consistency if you can have only one.

File: integration/embeddium/src/main/java/dev/nolij/zume/integration/implementation/embeddium/ZumeEmbeddiumConfigScreen.java
Comment (Nolij) [inline integration/embeddium/src/main/java/dev/nolij/zume/integration/implementation/embeddium/ZumeEmbeddiumConfigScreen.java:4]: `*` import begone

File: legacy/src/main/java/dev/nolij/zume/legacy/LegacyZume.java
Comment (Nolij) [inline legacy/src/main/java/dev/nolij/zume/legacy/LegacyZume.java:55]: Revert. Maybe use `{}` instead of `{ }` though for consistency.

File: neoforge/src/main/java/dev/nolij/zume/neoforge/NeoZume.java
Comment (Nolij) [inline neoforge/src/main/java/dev/nolij/zume/neoforge/NeoZume.java:19]: `*` import begone

File: neoforge/src/main/java/dev/nolij/zume/neoforge/NeoZume.java
Comment (Nolij) [inline neoforge/src/main/java/dev/nolij/zume/neoforge/NeoZume.java:116]: Remove the extra indentation, but keep the whitespace stripping at the end of the line.

File: settings.gradle.kts
Comment (Nolij) [inline settings.gradle.kts:8]: Is `version` not a function here?? Kotlin syntax never ceases to find new ways to disgust me...

File: src/main/java/dev/nolij/zume/FabricZumeBootstrapper.java
Comment (Nolij) [inline src/main/java/dev/nolij/zume/FabricZumeBootstrapper.java:15]: This feels like too much indentation. Maybe move the `"""` to the next line to avoid it?

File: src/main/java/dev/nolij/zume/ForgeZumeBootstrapper.java
Comment (Nolij) [inline src/main/java/dev/nolij/zume/ForgeZumeBootstrapper.java:23]: Same here.

File: src/main/java/dev/nolij/zume/ZumeMixinPlugin.java
Comment (Nolij) [inline src/main/java/dev/nolij/zume/ZumeMixinPlugin.java:42]: Revert (also applies to the methods below this one in this file).

**Review by Nolij — DISMISSED** [review]
Please keep in mind I only looked at the diff from `a92261e7ef320ba7375377b89da817d94517faee` when making this review.

**Comment by rhysdh540** [comment]
Sorry for wasting your reviewing time but I just reverted that commit

**Review by embeddedt — APPROVED** [review]

**Pull request merged**
